Restrict usage of WANT_READ to support event-driven IO

hanno-arm opened the pull request:
__Problem:__ The error code `MBEDTLS_ERR_SSL_WANT_READ` is overloaded with two different meanings:

1.  Firstly, it can indicate that there is no further data available from the underlying transport

2. Secondly, it can indicate that further processing of internal buffers is necessary and that the previously called function should be called again.

The lack of separation between these two meanings impedes the use of event-driven IO, because the user cannot know when to idle on the underlying transport and when to immediately call the previous function again. This was also the problem that triggered the issue in the first place in #772.

__This PR:__ Restricts the meaning of `MBEDTLS_SSL_WANT_READ` to case (1), looping in the cases where it was previously used for the purpose of (2). 

Further, it provides the following enhancements to using and testing event-driven I/O:

__Test infrastructure:__ 
* A `pack=TIME` option for our UDP proxy allowing to merge multiple records into a single UDP datagram.
* An 'UDP proxy wrapper' which silently puts a server binary like `ssl_server2` behind a UDP proxy. This is useful e.g. to run our test-suites with UDP packing added everywhere without the need to change server application or test script.
* Add `event=1` option to `ssl_server2` and `ssl_client2` to enable event-driven IO. With that, the applications idle when receiving a `WANT_READ` or `WANT_WRITE`, until they either receive an event from the timer or the underlying transport.

__Library:__
* Add a new public polling function for `mbedtls_net_context_t` allowing to wait for events on mbed TLS provided network contexts.
* Add a new public function `mbedtls_ssl_check_pending` to check whether the internal buffers still contain data to be processed. This is an inevitable addition because a user employing event driven IO needs that information to decide when he/she can idle on the underlying transport. Note that this is different from `mbedtls_ssl_get_bytes_avail` which only concerns application data. A simple usecase demonstrating the need for `mbedtls_ssl_check_pending` is the case when multiple application data records are contained in a single UDP datagram. Having fully read the data from the first datagram, `mbedtls_ssl_get_bytes_avail` will return `0`, but idling on the underlying transport might deadlock.

__References:__ Originally triggered by #772, internal reference IOTSSL-1204

@mpg Please review.

RonEld commented on library/ssl_srv.c:
```diff
@@ -3585,15 +3585,19 @@ static int ssl_parse_certificate_verify( mbedtls_ssl_context *ssl )
 
-        if( ( ret = mbedtls_ssl_read_record_layer( ssl ) ) != 0 )
+        do ret = mbedtls_ssl_read_record_layer( ssl );
+        while( ret == MBEDTLS_ERR_SSL_CONTINUE_PROCESSING );
```
Couldn't we get into and endless loop here?
What will happen if the peer socket gets disconnected, so f_recv will always return failure, causing:
~~~
     if( ( ret = mbedtls_ssl_fetch_input( ssl, mbedtls_ssl_hdr_len( ssl ) ) ) != 0 )
      {		      {
 @@ -3815,8 +3818,7 @@ int mbedtls_ssl_read_record_layer( mbedtls_ssl_context *ssl )
                                              "(header)" ) );		                                              "(header)" ) );
              }		              }
  		  
 -            /* Get next record */		 +            return( MBEDTLS_ERR_SSL_CONTINUE_PROCESSING );
 -            goto read_record_header;		
          }		          }
~~~
inside `mbedtls_ssl_read_record_layer`.

hanno-arm replied:
@RonEld The quote suggests that `return( MBEDTLS_ERR_SSL_CONTINUE_PROCESSING )` is within the failure branch of `mbedtls_ssl_fetch_input`, but it is within the failure branch for `mbedtls_ssl_parse_record`, which is only executed if the interface to the underlying transport `mbedtls_ssl_fetch_input` was successful. In case the latter fails, the error is escalated back to the user, and no looping takes place. 

The question of whether looping on unexpected records is desirable or not is a separate issue that is not touched upon by this PR (both the old and the suggested code do looping in some way) and is open to discussion - see IOTSSL-1424.

RonEld replied:
@hanno-arm you are correct, my mistake regarding the networking part. However, is it possible we will endless loop on invalid record? I mean, `mbedtls_ssl_read_record_layer` will call `mbedtls_ssl_parse_record` which will fail with `MBEDTLS_ERR_SSL_INVALID_RECORD` error to its caller (`mbedtls_ssl_read_record_layer`), which  will return  `MBEDTLS_ERR_SSL_CONTINUE_PROCESSING `.

hanno-arm replied:
@RonEld Yes exactly, that's possible and a potential source for DoS - IOTSSL-1424 was raised to discuss and potentially rework the corresponding code paths. This PR keeps things as they are in this regard, only implementing the loop slightly differently.

RonEld replied:
@hanno-arm but with current implementation, if `mbedtls_ssl_parse_record_header` fails, the error will be returned from `mbedtls_ssl_read_record_layer ` to `ssl_parse_certificate_verify`, which will fail and return error to user. Am I missing something?

RonEld replied:
I am talking about TLS , not DTLS connection

hanno-arm replied:
@RonEld I don't yet understand what you mean, could you elaborate? How does `parse_certificate_verify` enter?

RonEld replied:
@hanno-arm I missed that `MBEDTLS_ERR_SSL_CONTINUE_PROCESSING ` is returned by `mbedtls_ssl_read_record_layer ` only in case of DTLS, so instead of `goto read_record_header`, which is ugly looping, you do the loop at a one higher level. I retract my comment

hanno-arm replied:
@RonEld Yes exactly, for that particular part of the code the changes are meant to only change the way the looping is done, but essentially leaving the code as it was before. The main change is when we introduce looping where we previously returned `WANT_READ`. 
Thanks for looking at the code, and let me know if there are potential issues in the other parts of the PR.

hanno-arm commented:
@mpg Looping on `ssl->in_offt` in `mbedtls_ssl_read` is the most delicate part of these changes, as it functionally differs from repeated or recursive calls to `mbedtls_ssl_read` by omitting the checks at the beginning. This needs careful review, and I also want to give it a second thought. Don't merge yet, please.

RonEld commented on library/ssl_tls.c:
```diff
@@ -3897,46 +3899,6 @@ int mbedtls_ssl_read_record_layer( mbedtls_ssl_context *ssl )
-        }
-    }
-#endif
-
```
@hanno-arm why was this deleted?

RonEld commented on library/ssl_tls.c:
```diff
@@ -6618,7 +6589,9 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
-                    return( MBEDTLS_ERR_SSL_WANT_READ );
+                {
+                    continue;
+                }
```
@hanno-arm since this is DCP , packets can get lost. What will be the behavior in this case? The comment says "drop the packet", but now, we loop, instead of returning `MBEDTLS_ERR_SSL_WANT_READ` to the caller.

hanno-arm replied:
@RonEld The previous meaning of `WANT_READ` at this point was to signal to the user that he should call `mbedtls_ssl_read` again for further message processing. This is not what `WANT_READ` is supposed to express, and now we immediately do the fetching and processing of the next message within `mbedtls_ssl_read`. 

So in short, we're dropping / ignoring the packet in both cases, but in a slightly different way. 

The most direct way to functionally mimic a user calling `mbedtls_ssl_read` immediately again would be to do a recursive call or to jump to the beginning of the function. A recursive call is no option because we don't rely on tail-recursive optimizations by the compiler, and a goto jump to the beginning isn't good style. It seemed to me that a reasonble place to jump to for the next iteration would be the `if( ssl->in_offt == NULL)` check, but it needs to be double-checked that the part from the beginning of `mbedtls_ssl_read` skipped this way is indeed irrelevant for subsequent iterations.

RonEld replied:
@hanno-arm I see, in this case, do you think the sample applications should be modified? At the moment, they call `mbedtls_ssl_handshake`, if the error  is `WANT_READ` or `WANT_WRITE`. If now `WANT_READ` is not returned, will the applications work? Should they be modified not to check  `WANT_READ`?

hanno-arm replied:
@RonEld The previous loop-driven example code should be ok both for the previous and the modified code base; for example, in `ssl_client2.c` there are two places where `mbedtls_ssl_read` is called, and they are both tightly wrapped in a loop repeating as long as `WANT_READ` or `WANT_WRITE` is returned. If one does it like this, blurring the two meanings of `WANT_READ` goes unnoticed, because in either case it's ok to immediately try `mbedtls_ssl_read` again. With the modified code, `mbedtls_ssl_read` will return `WANT_READ` fewer times, and when it does, it means that the the user can idle on the underlying transport to signal pending data, i.e. to employ event-driven IO - something that was not possible, or at least very error-prone, before.

So the example applications should stay fine the way they are, but it would be good to add sample code showing how to correctly use event-driven IO. I've made some changes in https://github.com/ARMmbed/mbedtls/pull/942 in that direction.

RonEld commented on library/ssl_tls.c:
```diff
@@ -6631,7 +6604,9 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
-                    return( MBEDTLS_ERR_SSL_WANT_READ );
+                {
+                    continue;
+                }
```
same comment as before

RonEld commented on library/ssl_tls.c:
```diff
@@ -6720,7 +6702,7 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
         {
             MBEDTLS_SSL_DEBUG_MSG( 2, ( "ignoring non-fatal non-closure alert" ) );
-            return( MBEDTLS_ERR_SSL_WANT_READ );
+            continue;
```
This is not ignoring the non-fatal non-closure alert

hanno-arm replied:
@RonEld Can you elaborate? We _do_ ignore the message by going into the next iteration here, because the latter will fetch a new record. That's also what's happening when we return `WANT_READ` to the user and have him call `mbedtls_ssl_read` again.

RonEld replied:
ok

RonEld commented:
@hanno-arm I looked at other parts of the code, and as you said, instead of returning `MBEDTLS_ERR_SSL_WANT_READ`, we now loop.
My main concern is that on UDP (which DTLS runs on), we might enter endless loop, in case packets are lost

hanno-arm commented:
@RonEld I couldn't yet fully follow your concerns and have tried to explain the changes in more detail.

RonEld commented:
@hanno-arm Thanks. Obviously you investigated this much more than me. I just wanted to think of any corner cases. Please check that the sample application are working and not need to be modified as well

hanno-arm commented:
@RonEld Indeed, and I still owe you an explanation for the removal of the handshake flight retransmission code, which I will add later. @mpg asked about that, too.

hanno-arm commented:
@mpg Looping on `ssl->in_offt` in `mbedtls_ssl_read` is the most delicate part of these changes, as it functionally differs from repeated or recursive calls to `mbedtls_ssl_read` by omitting the checks at the beginning. This needs careful review, and I also want to give it a second thought. Don't merge yet, please.

sbutcher-arm commented:
@hanno-arm - This PR is marked as do not merge, and is still pending review by @mpg. Hanno, what work is remaining on this PR so I can merge it? Just @mpg's approval, or something else?

hanno-arm commented:
@sbutcher-arm The PR has already undergone detailed discussion on the approach with @mpg and @RonEld but is lacking in-depth review from @mpg; further, I marked it as _don't merge_ because I also wanted to give it yet another check myself as it touches the critical code-path around message-processing.

Progress deliberately came to halt when we decided to treat it as part of the message processing rework.

We can either revive this PR and its sibling #942 and attempt to merge it soon, in which case I expect it to be a matter of a week to get it rebased and reviewed given the previous discussions - or continue to wait for it to be part of the rework. The former comes at the cost of duplication of effort as the code-path will most likely be completely rewritten at some point, but at the benefit of finally making available better support for non-blocking I/O, which is pending for a long time. The rework being a longer-term effort, my recommendation would be to accept potential duplication and include this PR and #942 soon.

@mpg What are your thoughts?

mpg commented:
I fully agree with @hanno-arm's message.

I don't expect the review process to be very long or eventful either given the previous discussion; the main reason I haven't done it yet is I wasn't sure if this was still live. I support Hanno's conclusion that it's a trade-off between long-term duplication of work but fixing in a somewhat timely fashion an issue that is blocking for some users and that in that specific case I think it's better to fix it now than to wait for the long-term solution.

erja-gp commented:
@hanno-arm : can you also verify if the changes which you propose here will fix the issue which is described in: https://github.com/openthread/openthread/issues/2176

hanno-arm commented:
@erja-gp: Yes, I'll look into this, too, as soon I continue working on this issue.

hanno-arm commented:
@erja-gp Could you check if this PR fixes your problem?

hanno-arm commented:
retest CI

hanno-arm commented:
This is still missing a ChangeLog entry and giving credit to the author of #772.

hanno-arm commented:
retest CI

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2443,6 +2500,29 @@ int mbedtls_ssl_renegotiate( mbedtls_ssl_context *ssl );
  *                 again, or not transmitting the new identity to the
  *                 application layer, would allow authentication bypass!
+ *
+ *                 If this function returns something other than a positive
```
Yes, I agree, one should still be able to write something, but I don't think it's worth detailing on that in the documentation.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2443,6 +2500,29 @@ int mbedtls_ssl_renegotiate( mbedtls_ssl_context *ssl );
+ *                 MBEDTLS_ERR_SSL_CLIENT_RECONNECT, then the ssl context
+ *                 becomes unusable, and you should either free it or call
+ *                 \c mbedtls_ssl_session_reset() on it before re-using it for
+ *                 a new connection; the current connection must be closed.
```
That's pre-existing and was only moved. Looking at it, I don't like either formulation very much because they both read as if the user should actively do something about closing, which is not the case. I'd suggest to drop this half-sentence it altogether, because the paragraph already clearly states it is not allowed to do anything apart from `free` or `reset` with the context.

hanno-arm replied:
Removed the last half-sentence.

hanno-arm replied:
@mpg Thanks, I'll change it while we're at it.

hanno-arm replied:
@mpg Addressed in 298a7b2

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2235,11 +2236,54 @@ void mbedtls_ssl_conf_renegotiation_period( mbedtls_ssl_config *conf,
+ * \return         0 if nothing's pending, 1 otherwise.
+ *
+ * \note           This function is essential when using the library
+ *                 with event-driven I/O. The user should not idle
```
Done.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2354,6 +2398,19 @@ int mbedtls_ssl_get_session( const mbedtls_ssl_context *ssl, mbedtls_ssl_session
  *                 MBEDTLS_ERR_SSL_HELLO_VERIFY_REQUIRED (see below), or
  *                 a specific SSL error code.
  *
+ *                 If MBEDTLS_ERR_SSL_WANT_READ is returned, the handshake is
```
Done.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2354,6 +2398,19 @@ int mbedtls_ssl_get_session( const mbedtls_ssl_context *ssl, mbedtls_ssl_session
+ *                 and it is safe to idle until the timer or the underlying transport
+ *                 signal a new event. This is not true for a successful handshake,
+ *                 in which case the currently processed underlying transport's datagram
+ *                 might or might not contain further DTLS records.
```
Done.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2417,20 +2474,20 @@ int mbedtls_ssl_renegotiate( mbedtls_ssl_context *ssl );
- *                 MBEDTLS_ERR_SSL_CLIENT_RECONNECT (see below), or
- *                 another negative error code.
+ * \return         One of the following:
+ *                 - 0 for EOF, or
```
Done.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2443,6 +2500,29 @@ int mbedtls_ssl_renegotiate( mbedtls_ssl_context *ssl );
+ *                   from the underlying transport layer is currently being processed,
+ *                   and it is safe to idle until the timer or the underlying transport
+ *                   signal a new event.
+ *                 - If the function returns MBEDTLS_ERR_SSL_WANT_READ this does not mean
```
Done.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2443,6 +2500,29 @@ int mbedtls_ssl_renegotiate( mbedtls_ssl_context *ssl );
+ *                   request from the peer. Therefore, the user must be prepared to receive
+ *                   MBEDTLS_ERR_SSL_WANT_READ even when reacting to an incoming-data event
+ *                   from the underlying transport.
+ *                 - On success, the currently processed underlying transport's datagram
```
Done.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -6383,6 +6368,67 @@ size_t mbedtls_ssl_get_bytes_avail( const mbedtls_ssl_context *ssl )
+
+    if( ssl->keep_current_message == 1 )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "ssl_check_pending: record "
```
Done.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -6383,6 +6368,67 @@ size_t mbedtls_ssl_get_bytes_avail( const mbedtls_ssl_context *ssl )
+     */
+
+    /* TODO This needs correction in the same way as
+     *      read_record_layer, see IOTSSL-1414 */
```
I've removed the internal references now.

hanno-arm commented on library/ssl_tls.c:
```diff
@@ -6819,13 +6856,11 @@ int mbedtls_ssl_read( mbedtls_ssl_context *ssl, unsigned char *buf, size_t len )
      */
 
-    if( ssl->in_offt == NULL )
+    while( ssl->in_offt == NULL )
```
@gilles-peskine-arm `ssl->in_offt` is set if and only if an application data record is being read. I've added a comment.

Regarding outsourcing the loop, I'd rather keep changes minimal in this PR and leave improvements of structure to the larger scale reworkings of the message processing.

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -597,16 +608,22 @@ fi
 # sanity checks, avoid an avalanche of errors
-if [ ! -x "$P_SRV" ]; then
-    echo "Command '$P_SRV' is not an executable file"
+P_SRV_BIN=$(echo "$P_SRV" | sed -r -n "s/^([^ ]*).*$/\1/p")
```
Great!

hanno-arm commented on programs/test/udp_proxy_wrapper.sh:
```diff
@@ -0,0 +1,103 @@
+VERBOSE=1
+
+PARAM_SEP="^(.*)--(.*)$"
+PROXY_PARAMS=$(echo $@ | sed -n -r "s/$PARAM_SEP/\1/p")
```
I'd prefer to stick to spaces as separators; would you be ok with having ` -- ` (including spaces) as a separator to not accidentally take something like `--arg=val` as the separator (if this parameter form is ever used in our programs)? 

In any case, thanks for the explanation and the suggestion to switch `IFS` to newline - I didn't know about that.

hanno-arm commented:
@gilles-peskine-arm @RonEld Regarding: https://github.com/ARMmbed/mbedtls/commit/c76c619dd08eba3210caa6d13a8cca43e3a697fa. Apart from unification, it's a minor flaw that `ssl->in_msg[0]` is [read](https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L4013) before any check on `ssl->in_msglen` is performed. As `ssl->in_msg` always points to the beginning of the statically sized message buffer there's no danger of an overflow, but still a length-zero handshake message should be treated as an invalid record and not lead to retransmission. 

I'm not entirely sure if there was yet another reason, though.

gilles-peskine-arm commented:
@hanno-arm For `udp_proxy_wrapper.sh`: https://github.com/gilles-peskine-arm/mbedtls/blob/iotssl-1204-udp_proxy_wrapper/programs/test/udp_proxy_wrapper.sh

hanno-arm commented on library/ssl_srv.c:
```diff
@@ -3790,15 +3790,19 @@ static int ssl_parse_certificate_verify( mbedtls_ssl_context *ssl )
     /* Read the message without adding it to the checksum */
     do {
```
Here too I'd prefer not to change this because the code will anyway be thoroughly reworked soon.

hanno-arm commented:
@RonEld I've removed your initial approval because it applied to a by now very old version of the PR.

gilles-peskine-arm commented on programs/ssl/ssl_server2.c:
```diff
@@ -1328,26 +1422,37 @@ int main( int argc, char *argv[] )
+    if( opt.event == 1 && opt.nbio != 1 )
+    {
+        mbedtls_printf( "Warning: event-driven IO mandates nbio=1"
+                        " - overwrite\n" );
```
Please don't split debug messages.

hanno-arm commented:
@mpg I added some commits addressing your comments on #942. @gilles-peskine-arm Please have a look. 

Besides, unfortunately, something must still be broken in the process-killing: If I run
~~~
P_SRV="../programs/test/udp_proxy_wrapper.sh pack=100 -- event=1" P_CLI="../programs/ssl/ssl_client2 event=1" ./tests/ssl-opt.sh --filter "DTLS"
~~~
and abort after some time, I see multiple `ssl_server2` processes still running.

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2243,15 +2243,6 @@ void mbedtls_ssl_conf_renegotiation_period( mbedtls_ssl_config *conf,
  *
  * \return         0 if nothing's pending, 1 otherwise.
```
I suggest to make this “a positive value otherwise”, in case we want to encode what is pending in the return value in the future.

gilles-peskine-arm commented:
Note: error in `udp_proxy.c` on platforms where `size_t` and `unsigned long` are different (https://jenkins-internal.mbed.com/job/mbedtls-pr/414/execution/node/199/log/)

hanno-arm commented:
@mpg @gilles-peskine-arm Added commit removing the busy-waiting in `udp_proxy.c`.

gilles-peskine-arm commented:
I'm confused about the CI indication on https://github.com/ARMmbed/mbedtls/pull/946/commits/77abef5cba14ebd2f396e786615fb0e9de4b9338 which should have fixed the `size_t`/`unsigned long` mismatch. https://jenkins-internal.mbed.com/job/mbedtls-pr/415/ shows the same failure, it seems to have run on an earlier commit?

gilles-peskine-arm commented:
Approved, conditional to a successful CI run (excepting the intermittent timing-related failures we're aware of).

gilles-peskine-arm commented:
CI didn't actually run? https://jenkins-internal.mbed.com/job/mbedtls-pr/416/ Let's try again.

mpg commented on library/ssl_srv.c:
```diff
@@ -3790,15 +3790,19 @@ static int ssl_parse_certificate_verify( mbedtls_ssl_context *ssl )
     /* Read the message without adding it to the checksum */
```
Pre-existing, only for reference for future cleanups: I think we have a layering violation here, and we should re-arrange code to only ever call `mbedtls_ssl_read_record()` and never the lower-level `mbedtls_ssl_read_record_layer()+handle_message_type()` (which ideally would be made static when we get the chance to do so). I think this could be achieved simply by calling `ssl->handshake->calc_verify()` before we read the message, but this should be double-checked as maybe it relies on state that has been modified in the meantime.

hanno-arm replied:
@mpg I intend to unify this in the forthcoming reworkings by setting a flag for each message indicating whether it contributes to the checksum or not, and to do the checksum update at the _end_ of message processing by the logic layer.

mpg replied:
Sounds good!

mpg commented on ChangeLog:
```diff
@@ -1,5 +1,15 @@
 
+= mbed TLS x.x.x branch released xxxx-xx-xx
+
+API changes
```
Can we rename to "API extensions" of something else that reflects that the API remains compatible, as is our policy? I don't want to mislead people into being afraid of incompatibilities while we actually do the hard work of paying attention to compatibility.

gilles-peskine-arm replied:
I personally agree, but it's out of scope of this PR, we'll discuss it internally.

mpg commented on ChangeLog:
```diff
@@ -2,6 +2,15 @@ mbed TLS ChangeLog (Sorted per branch, date)
 = mbed TLS x.x.x branch released xxxx-xx-xx
 
+Bugfix
+   * Restrict usage of error code MBEDTLS_ERR_SSL_WANT_READ to situations
```
Not a blocker, but I think it's better to write ChangeLog entries from the perspective of the user first, and developer second if at all. IMO the important information for users is that we fix deadlock situations with event-driven I/O, and it's better if the user doesn't have to read a paragraph about what internal changes we made to fix that before getting to the bit that interests them.

mpg commented on programs/test/udp_proxy_wrapper.sh:
```diff
@@ -108,7 +108,7 @@ if [ $VERBOSE -gt 0 ]; then
 fi
 
-"$SRV_BIN" "$@" >&2 &
+exec "$SRV_BIN" "$@" >&2 &
```
I think you want to remove the following lines then, it doesn't make sense to keep unreachable statements around.

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -2418,10 +2418,10 @@ int mbedtls_ssl_get_session( const mbedtls_ssl_context *ssl, mbedtls_ssl_session
- *                 \c mbedtls_ssl_session_reset() on it before re-using it for
- *                 a new connection; the current connection must be closed.
+ *                 MBEDTLS_ERR_SSL_WANT_READ/WRITE, you must stop using
+ *                 the SSL context for reading or writing, and either free it or
```
As this is `mbedtls_ssl_handshake()` we should adapt the wording here, eg "you must not attempt to continue the handshake with this context, nor to use if for reading or writing, and..."

mpg commented on include/mbedtls/ssl.h:
```diff
@@ -2438,10 +2438,10 @@ int mbedtls_ssl_handshake( mbedtls_ssl_context *ssl );
- *                 \c mbedtls_ssl_session_reset() on it before re-using it for
- *                 a new connection; the current connection must be closed.
+ *                 MBEDTLS_ERR_SSL_WANT_READ/WRITE, you must stop using
+ *                 the SSL context for reading or writing, and either free it or
```
Same as above.

mpg commented:
Note the CI failure seems to be related to the CI software (or the github API), so can be disregarded. However, as the tests did not run, we probably want to run them manually before merging.

hanno-arm commented:
@mpg Some test cases do not work with `udp_proxy_wrapper.sh` because they rely on packet not being merged by Mbed TLS - that's why I'd be inclined to keep `udp_proxy_wrapper.sh` outside the CI infrastructure for now.

mpg commented:
@hanno-arm I'm afraid this reinforces my impression that I don't get what the point of this script it. However, that's not a blocker for me. The only real blocker for me is that test cases with `event=1` (without and with proxy duplicating-delaying-droping and/or packing) need to be added to `ssl-opt.sh`.

mpg commented:
Update: CI failed with compile errors: https://jenkins-internal.mbed.com/job/mbedtls-pr/418/execution/node/229/log/

hanno-arm commented:
@mpg The script is mainly for ad-hoc testing of specific situations and convenience in using the UDP proxy (it's almost as quick to start a server with a proxy as without). I also found it interesting to run `ssl-opt.sh` with it, the bug https://github.com/ARMmbed/mbedtls/pull/946/commits/52c6dc64c675d24b1be65e06d43dd67dc111e762 was found this way if I remember correctly - still, running this continuously is not possible because as mentioned some tests depend on the current non-merging behavior of Mbed TLS.

I'm not insisting on adding it to the repository, but I think it can be useful and convenient to have it, even if it's not an integral part of the CI.

mpg commented:
@hanno-arm Thanks for clarifying! I have nothing against having this script in the repo, on the contrary I agree it can be convenient for manual testing. I seem to have misundertood some earlier comment, which is why I thought it was for the CI.

hanno-arm commented:
@mpg I added tests for the `event` and `pack` options.

I haven't yet moved the code-path regarding `handshake_wrapup_hs_transform`, as the discussion is still pending.

hanno-arm commented:
Updated the PR description to also mention the changes from the now-closed PR #942.

gilles-peskine-arm commented:
There are build failures under Windows and under Linux-x86. @hanno-arm could you please investigate?

hanno-arm commented:
@gilles-peskine-arm Something is wrong with the CI, it still complains about an old sign issue which is resolved I think.

gilles-peskine-arm commented:
@hanno-arm Oh, I see, CI ran against an old commit again. I'll trigger a new build using a different job.

gilles-peskine-arm commented:
Running https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/34/ against commit 72a4f0338d08712209c909d7cdf9853d2cb4d3cf

hanno-arm commented:
@gilles-peskine-arm Pushed commit to fix sign conversion warning under MSVC.

gilles-peskine-arm commented:
Automatic CI picked an old commit again. Running manually: https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/35/

hanno-arm commented:
Started CI manually again: https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/39/

hanno-arm commented:
@gilles-peskine-arm Do you see any obstacles for this one?

gilles-peskine-arm commented:
There's still a build failure on Windows in `udp_proxy.c`: https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/39/execution/node/189/log/

hanno-arm commented:
@gilles-peskine-arm Thanks. Pushed a5e6897 to solve this. @gilles-peskine-arm @mpg please review.

hanno-arm commented:
Manually started CI: https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/41/

hanno-arm commented:
@gilles-peskine-arm CI successful up to timing, as far as I see.

hanno-arm commented:
@gilles-peskine-arm Any news on this one?

jwhui commented:
@gilles-peskine-arm, what is the status of this PR?

sbutcher-arm commented:
Hi @jwhui,

This is currently going through internal review, pending design approval as you can see from the labels, and is planned to be incorporated into a release in about 6 weeks time.

thcu-gp commented:
Are there any updates on when this will be included in an mbedtls release. Last update said "in +/- 6 weeks" and that is about 6 weeks ago. And there has been a request for changes less than two weeks ago.

We have some dependencies on this fix (or at leas the fix for #772)

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2235,11 +2236,59 @@ void mbedtls_ssl_conf_renegotiation_period( mbedtls_ssl_config *conf,
+ *                 that all internal data has been processed.
+ *
+ */
+int mbedtls_ssl_check_pending( const mbedtls_ssl_context *ssl );
```
@sbutcher-arm The use of this function is demonstrated in the new event-driven I/O code in [`ssl_server2`](https://github.com/hanno-arm/mbedtls/blob/a5e68979cabc0883935fafd3e5cc86418e6a4239/programs/ssl/ssl_server2.c#L883) and [`ssl_client2`](https://github.com/hanno-arm/mbedtls/blob/a5e68979cabc0883935fafd3e5cc86418e6a4239/programs/ssl/ssl_client2.c#L481).

hanno-arm replied:
@sbutcher-arm My apologies, my answer refers to `net_poll`, but not `check_pending`. I'll update shortly.

hanno-arm replied:
@sbutcher-arm Thanks for bringing this up, I extended `ssl_server2.c` and the test suites.

hanno-arm commented on library/net_sockets.c:
```diff
@@ -433,6 +433,58 @@ int mbedtls_net_set_nonblock( mbedtls_net_context *ctx )
+/*
+ * Check if data is available on the socket
+ */
+
```
@sbutcher-arm This hasn't been investigated so far. Is it a question out of mere interest or do you see it blocking the PR?

sbutcher-arm replied:
Sadly, this is a blocking issue for me. I need to know either how it can be implemented in Mbed OS, or why I shouldn't care. Without knowing that, I can't approve it.

hanno-arm replied:
@sbutcher-arm Event-driven I/O is only possible if there is a means to idle a socket until a condition like availability of incoming data, or availability to send outgoing data, is met. The new function shows how this can be implemented on top of Unix or Windows sockets. It has no impact on the usability of Mbed TLS with Mbed OS: Both before and after this PR, use of event-driven I/O in Mbed OS requires some means of polling on the underlying socket.

sbutcher-arm replied:
Okay - that satisfies me. Thanks for the explanation @hanno-arm.

hanno-arm commented on ChangeLog:
```diff
@@ -1,5 +1,24 @@
+     returned when unexpected messages were being discarded, ignoring that
+     further messages could potentially already be pending to be processed
+     in the internal buffers; these cases lead to deadlocks in case
+     event-driven I/O was used. Found by Hubert Mis.
```
Thanks, fixed in 62dcbaf.

hanno-arm commented:
@sbutcher-arm Thanks for your review, I have left comments and pushed a commit to improve the ChangeLog.

hanno-arm commented:
@sbutcher-arm @gilles-peskine-arm @mpg I pushed 3 commits extending `ssl-opt.sh` and slightly modifying `ssl_server2` to demonstrate the use of `mbedtls_ssl_check_pending`. In more detail:
* 6a33f59 Duplicates the existing event-driven I/O tests (that were TLS-only) for DTLS.
* ddc3ebb Modifies `ssl_server2.c` to show the use of `mbedtls_ssl_check_pending`.
* bc6c110 Adds a test case to `ssl-opt.sh` which triggers the scenario mentioned in [the documentation of `mbedtls_ssl_check_pending`](https://github.com/hanno-arm/mbedtls/blob/a5e68979cabc0883935fafd3e5cc86418e6a4239/include/mbedtls/ssl.h#L2262) which fails without the proper use of `mbedtls_ssl_check_pending` added in ddc3ebb (If you uncomment the `mbedtls_ssl_check_pending( &ssl ) == 0 && ...` part [here](https://github.com/hanno-arm/mbedtls/blob/bc6c1101399a31058cf93b3f31be65a8d11e4bb2/programs/ssl/ssl_server2.c#L2432) the test will deadlock as expected).

@mpg @gilles-peskine-arm Please review.

gilles-peskine-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -602,7 +661,8 @@ int main( int argc, char *argv[] )
             opt.request_size = atoi( q );
-            if( opt.request_size < 0 || opt.request_size > MBEDTLS_SSL_MAX_CONTENT_LEN )
+            if( opt.request_size < 0 ||
+                opt.request_size > MBEDTLS_SSL_MAX_CONTENT_LEN )
```
Could you please revert this whitespace change? There's been a semantic change in `development`, and removing this whitespace change would make the file mergeable.

hanno-arm replied:
Done in b6f880b.

Patater commented:
After merging with latest development, this fails `cmake -D CMAKE_BUILD_TYPE:String=Check . && make clean && make && ./tests/ssl-opt.sh --memcheck`. Please investigate.

hanno-arm commented:
@Patater Thanks for bringing this up. The failure is introduced in commit https://github.com/ARMmbed/mbedtls/pull/946/commits/d82d84664af1da94aef06febf860418f82f97358 - terminating the server via a `KILL` signal seems to interfere with valgrind @gilles-peskine-arm Any input from your side?

gilles-peskine-arm commented:
I hadn't thought about it. Now that you point it out, it's obvious: SIGKILL doesn't give valgrind the opportunity to do its exit checks. Try SIGQUIT instead?

hanno-arm commented:
@gilles-peskine-arm Thanks for the input. I am wondering, though, why the initial `SIGTERM` doesn't succeed in terminating the process. @Patater suggested it might be related to the termination taking longer with valgrind.

gilles-peskine-arm commented:
I guess that's the reason for d82d84664af1da94aef06febf860418f82f97358.

hanno-arm commented:
@gilles-peskine-arm The reason was that the sigterm wouldn't kill the server process if the latter was polling on a socket afaik.

hanno-arm commented:
@Patater @gilles-peskine-arm Pushed a commit changing the termination to use a sigquit instead of sigkill, and started a Jenkins run at https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/97/.

mpg commented:
@hanno-arm @gilles-peskine-arm @Patater I'm not sure just switching from `KILL` to `QUIT` is going to fix the problem. You still need the server to catch the `TERM`, goto `exit` and call the various `xxx_free()` functions otherwise valgrind is rightfully going to report memory leaks all over the place.

I think there are two sides to this issue:
1. The server needs to catch `SIGTERM` and do its cleanup before it exits.
2. Valgrind needs to do its thing before it exits.
I don't think addressing point 2 alone is going to be enough. I'm also not sure why we think 0.01s is always going to be enough for the server to do its cleanup (on a loaded machine under valgrind).

hanno-arm commented:
I pushed another commit increasing the time the UDP proxy accumulates records before dispatching a datagram from 10ms to 50ms, as 10ms wasn't enough for some tests when run with valgrind.

mpg commented:
I thought a bit more about it, and I really think `kill; wait` was correct an should not be changed. The programs (server, udp proxy) are supposed to react to `SIGTERM` by cleaning up and exiting. If they don't, it's a bug and we want to know about it and fix it rather than work around it with `sleep $arbitrary_delay; kill $some_other_signal`. The correct delay is the time that `wait` takes to return. If we want to have some timeout as a defense, fine, but reaching the timeout should be treated as an error (just the same way as when the client times out).

gilles-peskine-arm commented:
@mpg I agree with your reasoning. I think I was the one who suggested SIGKILL and now I'm trying to remember why. A 0.01s timeout does not make sense if the goal was a fallback in case the process goes into an infinite loop.

hanno-arm commented:
@mpg has spotted the crucial point: The TERM handler in `ssl_server2.c` invalidates the sockets and expects the remainder of the program to perform a graceful shutdown as a consequence of this, but the new `idle()`-loop doesn't react to socket closure. 

@mpg @gilles-peskine-arm I have pushed another commit fixing the idling loop and reverting the changes in `ssl-opt.sh`, going back to killing the server via a plain `kill; wait`. 

Please review.

hanno-arm commented:
Jenkins started at https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/98/.

gilles-peskine-arm commented on programs/ssl/ssl_client2.c:
```diff
@@ -444,16 +444,17 @@ static int ssl_sig_hashes_for_test[] = {
            mbedtls_timing_delay_context *timer,
            int idle_reason )
 {
 #endif
```
Could you please move the brace outside the conditional compilation? As it is, the unbalanced braces screw up navigation in editors. Likewise in `ssl_server2.c`.

hanno-arm replied:
Fixed in 9b2b66e.

hanno-arm commented:
Release job started at https://jenkins-internal.mbed.com/job/mbedtls-release/227/

gilles-peskine-arm commented on library/net_sockets.c:
```diff
@@ -433,6 +433,58 @@ int mbedtls_net_set_nonblock( mbedtls_net_context *ctx )
+    ret = select( fd + 1, &read_fds, &write_fds, NULL,
+                  timeout == (uint32_t) -1 ? NULL : &tv );
+
+    if( ret < 0 )
```
Shouldn't we restart the `select` call if it returns `EINTR`? On an `EINTR` return, the signal handler has been executed, but there may be nothing ready on the file descriptors, you need to call `select` again to find out.

hanno-arm replied:
Thanks for spotting! Fixed in 9ac6403.

gilles-peskine-arm replied:
That `WSAEINTR` fix is needed, but it sure makes the code hard to read. How about `#define MBEDTLS_EINTR` (in this file, not in a header) in the `#ifdef`, and keep a simple `while( ret == MBEDTLS_EINTR )` here?

hanno-arm replied:
Improved in ef52797.

hanno-arm commented:
@mpg @gilles-peskine-arm Pushed two commits addressing @gilles-peskine-arm's feedback.

hanno-arm commented:
There is a build problem on Windows because of the use of `EINTR`.

hanno-arm commented:
@mpg @gilles-peskine-arm Pushed another commit to fix a build failure under Windows where `EINTR` is not defined.

gilles-peskine-arm commented on library/error.c:
```diff
@@ -439,6 +439,8 @@ void mbedtls_strerror( int ret, char *buf, size_t buflen )
         if( use_ret == -(MBEDTLS_ERR_SSL_INVALID_VERIFY_HASH) )
             mbedtls_snprintf( buf, buflen, "SSL - Couldn't set the hash for verifying CertificateVerify" );
+        if( use_ret == -(MBEDTLS_ERR_SSL_CONTINUE_PROCESSING) )
+            mbedtls_snprintf( buf, buflen, "SSL - Internal-only message signalling that further message-processing should be done" );
```
“Signaling” and “signalling” are both correct, but pick one: `error.c` and `ssl.h` are not consistent. You need to re-run `scripts/generate_errors.pl`.

gilles-peskine-arm replied:
“Signaling” and “signalling” are both correct, but pick one: `error.c` and `ssl.h` are not consistent. You need to re-run `scripts/generate_errors.pl`.

gilles-peskine-arm commented on library/net_sockets.c:
```diff
@@ -45,6 +45,8 @@
 #if (defined(_WIN32) || defined(_WIN32_WCE)) && !defined(EFIX64) && \
     !defined(EFI32)
 
+#define MBEDTLS_EINTR WSAEINTR
```
Unfortunately `check-names.sh` rejects `MBEDTLS_EINTR` because it reserves the `MBEDTLS_` prefix for macros defines in `config.h`. I propose https://github.com/gilles-peskine-arm/mbedtls/commit/a1a16fed4fa8c0375dfd91c32002f69b893ee81a

mpg replied:
Note: not just `config.h`, any header file (as opposed to C files).

hanno-arm replied:
@gilles-peskine-arm Thanks for spotting that and for providing a fix. I'm ok with you pushing it to this branch.

gilles-peskine-arm replied:
I cherry-picked that commit onto your branch.

gilles-peskine-arm commented on library/error.c:
```diff
@@ -439,6 +439,8 @@ void mbedtls_strerror( int ret, char *buf, size_t buflen )
         if( use_ret == -(MBEDTLS_ERR_SSL_INVALID_VERIFY_HASH) )
             mbedtls_snprintf( buf, buflen, "SSL - Couldn't set the hash for verifying CertificateVerify" );
+        if( use_ret == -(MBEDTLS_ERR_SSL_CONTINUE_PROCESSING) )
+            mbedtls_snprintf( buf, buflen, "SSL - Internal-only message signalling that further message-processing should be done" );
```
“Signaling” and “signalling” are both correct, but pick one: `error.c` and `ssl.h` are not consistent. You need to re-run `scripts/generate_errors.pl`.

hanno-arm replied:
@gilles-peskine-arm Thanks, fixed in c9f46d.

gilles-peskine-arm commented:
I started test runs on a merge of this PR with development plus two bug fixes (https://github.com/gilles-peskine-arm/mbedtls/commits/pr_946%2Bdevelopment):

* https://jenkins-internal.mbed.com/job/mbedtls-release/240/
* https://jenkins-internal.mbed.com/job/mbedtls-commit-tests/99/

hanno-arm commented:
Failing to reproduce the CI failures so far - any input is welcome. A new run for comparison has been started at https://jenkins-internal.mbed.com/job/mbedtls-release/307.

hanno-arm commented:
Merged `development` into this branch, which hopefully resolves the CI issues.

@gilles-peskine-arm Do you have time to review the merge?

hanno-arm commented:
Thanks @k-stachowiak for checking the merge commit.

The event-driven IO tests in `ssl-opt.sh` are still failing under `MemSan` and `clang` for reasons I don't yet understand - I'm currently investigating this.

hanno-arm commented:
__Update:__ The reason for the CI failure with MemSan is that `FD_ZERO` is sometimes implemented in assembly, which is not compatible with MemSan.

hanno-arm commented:
Release job started: https://jenkins-internal.mbed.com/job/mbedtls-release/337/

gilles-peskine-arm commented on ChangeLog:
```diff
@@ -2,7 +2,30 @@ mbed TLS ChangeLog (Sorted per branch, date)
+     trusted CA with a non DER-compliant certificate. Found by luocm on GitHub.
+     Fixes #825.
+
+API changes
```
For better or for worse, we spell it “API Changes” these days.

gilles-peskine-arm commented on ChangeLog:
Merge glitch.
